tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,309 @@
+        """Instantiate this data set.
+
+        Args:
+            nspins: Python `int` number of spins in the system.
+            system_name: Python `str` name of the system.
+            data_dir: Python `str` location where to store the data on disk.
+             Default is `~/tfq-datasets`
+        """
+
+        # TODO: How should downloading and storing the data be handled?
+        if data_dir is None:
+            data_dir = Path.expanduser(Path('~/tfq-datasets'))
+            data_dir.mkdir(parents=True, exist_ok=True)
+        else:
+            data_dir = Path(data_dir)
+            assert data_dir.exists(), "{} does not exist".format(data_dir)
+        data_path = data_dir / Path(system_name + "_{}/".format(int(nspins)))
+        self.nspins = nspins
+        self._download_dataset(data_path)
+        # TODO: This only works for a single order parameter. For
+        #  the XY model or XYZ model, this will have to be generalized.
+        self.order_parameters, self.additional_info = \
+            self._get_spin_systems(data_path)
+
+    def __str__(self):
+        """String representation of SpinSystemDataset object"""
+        # TODO: This only works for a single order parameter. For
+        #  the XY model or XYZ model, this will have to be generalized.
+        return "System with {} order parameters in range [{},{}]".format(
+            len(self), self.order_parameters[0], self.order_parameters[-1])
+
+    def __len__(self):
+        """Length of this object is determined by number of order parameters"""
+        return len(self.order_parameters)
+
+    @staticmethod
+    def _get_spin_systems(path):
+        """Iterate over path and create `SpinSystem` for each order parameter"""
+        order_parameters = []
+        additional_info = []
+        SpinSystemInfo = namedtuple('SpinSystemInfo', [
+            'g', 'gs', 'gs_energy', 'res_energy', 'fidelity', 'data_path',
+            'params'
+        ])
+
+        for directory in [x for x in path.iterdir() if x.is_dir()]:
+            g = float(str(directory.name))
+            with open(directory / Path('stats.txt'), 'r') as file:
+                lines = file.readlines()
+                res_e = float(lines[0].split('=')[1].strip('\n'))
+                fidelity = float(lines[2].split('=')[1].strip('\n'))
+            order_parameters.append(g)
+            additional_info.append(
+                SpinSystemInfo(g=g,
+                               gs=np.load(directory / Path('groundstate.npy')),
+                               gs_energy=np.load(directory /
+                                                 Path('energy.npy'))[0],
+                               res_energy=res_e,
+                               fidelity=fidelity,
+                               data_path=str(directory),
+                               params=np.load(directory / Path('params.npy'))))
+
+        return list(zip(*sorted(zip(order_parameters, additional_info))))
```

Comment:
Why does the returned dataset need to be sorted ? Does this break the ordering with the original circuits returned by tfi_chain ?

Reply:
Without the sort, `order_parameters` looks like this:
```python
[1.26, 1.7, 0.88, 1.22, 1.32, 0.9, 0.52, ...., 1.16, 0.5, 1.6, 0.4, 1.64, 0.38]
```
due to the way the folder are read. Sorting here does not break the resolved circuits from `tfi_chain`, since we loop over the sorted `order_parameters[i]` and `additional_info[i]`
```python
for i in range(len(order_parameters)):
    parameter_map = dict(
        zip(symbol_names.flatten(),
            additional_info[i].params.flatten()))
    resolved_circuits.append(cirq.resolve_parameters(
        circuit, parameter_map))
    hamiltonians.append(get_hamiltonian(qubits,
                                        order_parameters[i]))
```
I agree that it is maybe a bit hacky